Examples/clim-fig.lisp

```diff
@@ -376,35 +402,65 @@
-  (let ((latest-undo-entry (pop (clim-fig-undo-list *application-frame*))))
+   'move an object', or the CLEAR command.
+
+   In the first case, remove the record and add it to the redo list;
+   in the second case, move the object back to its previous position;
+   to undo a CLEAR, replay the output-history."
+  (alexandria:when-let ((latest-undo-entry (pop (clim-fig-undo-list *application-frame*))))
     (cond
-      ((not latest-undo-entry) (beep))
+      ((typep latest-undo-entry 'clim-fig-move-event)
+       (multiple-value-bind (x y) (output-record-position (record latest-undo-entry))
+         (setf (output-record-position (record latest-undo-entry))
+               (values (- x (delta-x latest-undo-entry))
+                       (- y (delta-y latest-undo-entry))))
+         (window-refresh *standard-output*))
+       (push latest-undo-entry (clim-fig-redo-list *application-frame*))
+       (activate-gadget (find-pane-named *application-frame* 'redo)))
       ((listp latest-undo-entry)
        (loop for record in latest-undo-entry do
             (with-output-as-presentation (*standard-output* nil 'figure)
               (stream-add-output-record *standard-output* record))
-            (replay record *standard-output* (bounding-rectangle record))))
+            (replay record *standard-output* (bounding-rectangle record)))
+       (activate-gadget (find-pane-named *application-frame* 'clear))
+       (deactivate-gadget (find-pane-named *application-frame* 'redo)))
       (T
        (erase-output-record latest-undo-entry *standard-output*)
-       (push latest-undo-entry (clim-fig-redo-list *application-frame*))))))
+       (push latest-undo-entry (clim-fig-redo-list *application-frame*))
+       (activate-gadget (find-pane-named *application-frame* 'clear))))
+    (when (null (clim-fig-undo-list *application-frame*))
+      (deactivate-gadget (find-pane-named *application-frame* 'undo))
+      (deactivate-gadget (find-pane-named *application-frame* 'clear)))))
 
 (define-clim-fig-command com-redo ()
-  (alexandria:if-let ((record (pop (clim-fig-redo-list *application-frame*))))
-    (progn
-      (push record (clim-fig-undo-list *application-frame*))
-      (with-output-as-presentation (*standard-output* nil 'figure)
-        (stream-add-output-record *standard-output* record))
-      (replay record *standard-output* (bounding-rectangle record)))
-    (beep)))
+  (alexandria:when-let ((current-redo-entry (pop (clim-fig-redo-list
+                                                *application-frame*))))
+    (push current-redo-entry (clim-fig-undo-list *application-frame*))
+    (activate-gadget (find-pane-named *application-frame* 'undo))
+    (activate-gadget (find-pane-named *application-frame* 'clear))
+    (cond
+      ((typep current-redo-entry 'clim-fig-move-event)
+       (multiple-value-bind (x y) (output-record-position
+                                   (record current-redo-entry))
```

Comment:
put (output-record-position (record current-redo-entry)) in its own line (without break in function call)